File: March 2024 GFG SOLUTION/March-09.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // Function to find the nth character in the modified string.
    // Parameters:
    //   s: The original string.
    //   r: The number of iterations to perform the modification.
    //   n: The index of the character to return in the modified string.
    // Returns:
    //   The nth character in the modified string after r iterations.
    char nthCharacter(string s, int r, int n) {
        int len = s.length();  // Length of the original string.
        
        // Perform r iterations.
        for (int i = 0; i < r; i++) {
            string temp = s;  // Create a temporary string to hold the modified version.
            
            // Iterate through each character in the string.
            for (int j = 0; j < len; j++) {
                // Modify the character based on its position and the original value.
                if (temp[j / 2] == '0')
                    s[j] = '0' + (j % 2);
                else
                    s[j] = '1' - (j % 2);
            }
        }
        
        // Return the nth character in the modified string.
        return s[n];
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int R, N;
        string S;
        cin >> S >> R >> N;
        Solution ob;
        cout << ob.nthCharacter(S, R, N) << endl;
    }
    return 0;
}
// } Driver Code Ends

File: March 2024 GFG SOLUTION/March-09.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;


class GfG
{
    public static void main(String args[])
        {
            Scanner sc = new Scanner(System.in);
            int t = sc.nextInt();
            while(t-->0)
                {
                    String S = sc.next();
                    int R = sc.nextInt();
                    int N = sc.nextInt();
                    Solution obj = new Solution();
                    System.out.println(obj.nthCharacter(S,R,N));
                }
                
        }
}

// } Driver Code Ends


//User function Template for Java

class Solution {
    char nthCharacter(String s, int r, int n) {
        n += 1;
        StringBuilder f = new StringBuilder();
        int it = 0;
        int l = 0;

        while (l < n) {
            it += 1;
            l = it * (int)Math.pow(2, r);
        }

        int p = (int)Math.pow(2, r);

        n -= 1;
        n %= p;

        if (it == 0) {
            it += 1;
        }

        f.append(s.charAt(it - 1));

        for (int i = 0; i < r; i++) {
            StringBuilder sb = new StringBuilder();
            for (int j = 0; j < f.length(); j++) {
                if (f.charAt(j) == '1') {
                    sb.append("10");
                } else {
                    sb.append("01");
                }
            }
            f = sb;
        }

        return f.charAt(n);
    }

}
